### ISSUE TITLE
Hosting app in shinyapps.io

### ISSUE THREAD

**Issue by xhdong-umd** [ctmm-initiative/ctmm-webapp#19]
Before we explore the option of hosting the app by ourselves, we can use the service provided in shinyapps.io. The free plan is very limited so we need to at least consider [the basic plan $39/month](https://www.shinyapps.io/).

Some terms used:
- instance. I assume it be an Amazon EC2 virtual machine instance. The RAM of the instance range from 1G to 8G. Free plan can only have 1G instance, basic plan can have 8G instance.
- worker. One instance can start multiple R processes to support the app. Free plan can only have 1 worker per instance.
- connection. A browser connection between user and server. 

R is single threaded, this means one R process can only handle one request at a time. For simple app, because most operations can be finished within hundreds of ms, one worker can handle multiple users/connections without obvious latency problem. A worker can take 50 connections by default, but the server will start new worker to balance the load after 5% of the limit is reached, if there are quota for new worker.

According to my observation, our app use 200M memory for one worker (the first simple version of app only use 50M). So from memory aspect, 1G instance can have 3-4 worker, 8G instance can have 24 - 32 worker.

- If user only use app for small dataset and plots, assume the most expensive operations take <2 seconds, then 1 worker may handle 5 - 10 concurrent connections.
- If user used big dataset or operations that could take minutes, 1 worker can only handle one connection. 
  I did a simple test: With the free plan of 1G instance, 1 worker, I started a connection to use a modeling process that take minutes to finish. At the same time I tried to connect to server from another browser, and I found the connection cannot be established. Only after the modeling process finished, the 2nd browser is connected.

So my estimate is:
- **For simple tasks of plot etc**
  + **free plan can support 5 - 10 concurrent connections.**
  + **basic plan can support 40 - 80 concurrent connections.**
- **For complex tasks that take minutes to finish**
  + **free plan can support 1 concurrent connections.** The 1G instance still have potential here, but the plan limited the works it can launch.
  + **basic plan can support 24 - 32 concurrent connections.**

This is just an optimistic estimate, the real number probably will be lower especially when we want small latency. In the other hand, if we don't have a lot of users modeling at the same time, it may still serve the purpose.

The plan sign up is relatively simple, can be upgrade/downgrade at any time. Though it seemed that there is no discount for academic, and there is no partial refund if cancelled (if downgraded to a lower paid plan, the remaining credit can be kept).

**Comment by jmcalabrese** [comment]
@xhdong-umd, thanks for looking into that. 

You say the plan can be upgraded/downgraded at any time. Does that mean that we could sign up for the basic (paid) plan for one month for testing purposes, and then downgrade to the free plan after that month? Are you sure there are no penalties or additional fees for doing that?

I will look into how we can pay for the basic plan.

**Comment by xhdong-umd** [comment]
@jmcalabrese I think so. [Here is the official guide](http://docs.rstudio.com/shinyapps.io/billing-and-account-management.html). You can also sign up a free account to look at the actual page of billing. 

> All plans are payable by credit card only. 
> 
> All subscriptions auto-renew at the end of the subscription cycle. If you would to like subscribe for only one month or one year, you will need to cancel your subscription prior to the start of the next cycle.
> 
> A detailed PDF invoice is available for download on the Account -> Billing page.
> 
> An account can be upgraded or downgraded at any time. To change your plan, visit the dashboard under Account -> Billing, scroll to the Plan section, and select the plan you prefer. Upgrades take effect immediately
> 
> If you would like to cancel your paid subscription and return to a free account, you can downgrade to the Free plan within the Account -> Billing page.
> 
> Please note that at this time cancellations take effect immediately and the entitlements in the subscription cycle do not continue until the end of the subscription cycle. In addition, there are no partial refunds.

The account can be renamed (which determine how the url will look like, my current url is ctmm.shinyapps.io/appname), or transferred to somebody else. So you can pay for an account then transfer it to me if that make it easier.

**Comment by xhdong-umd** [comment]
There is not much need to be done here before actual sign up. I think we can sign up the basic plan sometime in August, and reserve several days for testing and adjusting configurations. 

I'm closing the issue temporarily now, and I will reopen it later.

**Comment by xhdong-umd** [comment]
We have signed up for the basic plan.

After some time tuning the app configurations and experimenting, now the hosted app can support several users using the time consuming model fitting feature at the same time. 

In theory our plan should support 
- up to 20 users for these kind of computation expensive procedures, this is less than my expectation of 32 because the parameter combinations are limited.
- up to 400 users for light operations

The actual limit is probably lower than these theoretical numbers.

I tested with 5 different connections using the model fitting at the same time without problem.
